### ISSUE TITLE
question about usage of std::equal in various places

### ISSUE THREAD

**Issue by RallyTronics** [sim642/calculate#2]
Sorry, I know this in an old project.  

When I try to run it in Visual Studio I get an exception in various places where std::equal was used if the strings are not the same length ( for example in the funcname_compare function when comparing "abs" to "+" ).  

Exception: cannot seek string iterator after end

```
struct funcname_compare
{
	bool operator() (const std::string &lhs, const std::string &rhs) const
	{
		if (equal(lhs.begin(), lhs.end(), rhs.begin())) // lhs is prefix of rhs
			return false;
		else if (equal(rhs.begin(), rhs.end(), lhs.begin())) // rhs is prefix of lhs
			return true;
		else
			return lhs < rhs;
	}
};

```

What is the expected behavior in the case when the strings are not the same length?

thanks!

**Comment by sim642** [comment]
Now looking at it, this seems to be really a out-of-bounds bug which GCC's standard library at the time just ignored I suppose. The two `std::equal` checks should probably have length comparisons before them indeed.

**Comment by RallyTronics** [comment]
I think it needs the version that takes both iterators for each

i.e.  equal(rhs.begin(), rhs.end(), lhs.begin(), lhs.end())

**Comment by sim642** [comment]
No, I think that would be incorrect because it would just check `lhs == rhs` but it needs to check if one is the prefix of the other or vice versa.
